src/app/services/geo-data.service.ts

```diff
@@ -298,10 +306,138 @@ export class GeoDataService {
+          this._tileServers.next(results);
+        });
+      }, (error => {
+        console.log(error);
+      }));
+  }
+
+  public updateTileServers(projectId: number, tileServers: Array<TileServer>) {
+    this.http.put(environment.apiUrl + `/projects/${projectId}/tile-servers/`, tileServers)
+      .subscribe( (resp) => {
+        this.getTileServers(projectId);
+      });
+  }
+
+  updateTileServer(projectId: number, tileServer: TileServer): void {
+    this.http.put(environment.apiUrl + `/projects/${projectId}/tile-servers/${tileServer.id}/`, tileServer)
+      .subscribe( (resp) => {
+        this.getTileServers(projectId);
+      });
+  }
+
+  public toggleTileServer(projectId: number, ts: TileServer) {
+    ts.uiOptions.isActive = !ts.uiOptions.isActive;
+    this.updateTileServer(projectId, ts);
+  }
+
+  getTileServers(projectId: number): void {
+    this.http.get(environment.apiUrl + `/projects/${projectId}/tile-servers/`).subscribe((tsv: Array<TileServer>) => {
+      tsv.sort((a, b) => {
+        return a.uiOptions.zIndex - b.uiOptions.zIndex;
+      });
+
+      console.log(tsv);
+
+      this._tileServers.next(tsv);
+    });
+  }
+
+  addTileServer(projectId: number, tileServer: TileServer) {
+    this.tileServers$.pipe(take(1)).subscribe(tileServerList => {
+      if (tileServerList) {
+        // TODO: Figure out a better way to handle ZIndex
+        let zIndexMax = -(tileServerList.length + 1);
+        tileServerList.forEach(ts => {
+          ts.uiOptions.zIndex = zIndexMax;
+          zIndexMax++;
+        });
+        this.updateTileServers(projectId, tileServerList);
+      }
+
+      tileServer.uiOptions.zIndex = -1;
+    });
+
+    this.http.post(environment.apiUrl + `/projects/${projectId}/tile-servers/`, tileServer)
+      .subscribe((resp) => {
+        this.getTileServers(projectId);
+      });
+  }
+
+  searchQMS(query: string, queryOptions: any): void {
+    const url = "https://qms.nextgis.com/api/v1/geoservices/";
+    const request = url +
+      "?search=" + query +
+      "&type=" + queryOptions['type'] +
+      "&ordering=" + queryOptions['order'] +
+      queryOptions['ordering'] +
+      "&cumulative_status=works";
+
+    this.http.get(request).subscribe((q) => {
+      this._qmsSearchResults.next(q);
+    });
+  }
+
+  getQMSTileServer(projectId: number, id: number) {
+    const request = "https://qms.nextgis.com/api/v1/geoservices/" + id;
+    this.http.get(request).subscribe((q) => {
+      const newServer: TileServer = {
+        name: q['name'],
+        type: q['type'],
+        url: q['url'],
+        attribution: q['desc'],
+        tileOptions: {
+          maxZoom: q['z_max'] ? q['z_max'] : null,
+          minZoom: q['z_min'] ? q['z_min'] : null,
+          layers: q['layers'] ? q['layers'] : null,
+          params: q['params'] ? q['params'] : null,
+          format: q['format'] ? q['format'] : null
+        },
+        uiOptions: {
+          opacity: 0.5,
+          isActive: false
```

Comment:
would this be true?

Reply:
These are leaflet layer settings. The tiles from QMS don't specify anything.
I changed it to be toggled on by default. Should I lower the opacity?

Reply:
> I changed it to be toggled on by default
👍 
>Should I lower the opacity?
hmm. I don't know. whatever you decide is good with me on this.